### PR TITLE
feat(Player): introduce `PlayerPacketHandlerChangeEvent`

### PR DESCRIPTION
**Pull request by cooldogedev**
### Behavioural changes

`NetworkSession->setHandler(PacketHandler)` now fires `PlayerPacketHandlerChangeEvent`

### CODE REVIEW

**Comment by dktapps** [comment]
What's the use case?

File: src/event/player/PlayerPacketHandlerChangeEvent.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+	public function getNetworkSession() : NetworkSession{
+		return $this->session;
+	}
+
+	public function getHandler() : ?PacketHandler{
+		return $this->handler;
+	}
+
+	public function setHandler(?PacketHandler $handler) : void{
```
Comment (dktapps) [inline src/event/player/PlayerPacketHandlerChangeEvent.php:41]: I don't like this. This opens lots of options to break things.

File: src/event/player/PlayerPacketHandlerChangeEvent.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+declare(strict_types=1);
+
+namespace pocketmine\event\player;
+
+use pocketmine\event\Event;
+use pocketmine\network\mcpe\handler\PacketHandler;
+use pocketmine\network\mcpe\NetworkSession;
+
+class PlayerPacketHandlerChangeEvent extends Event{
```
Comment (dktapps) [inline src/event/player/PlayerPacketHandlerChangeEvent.php:30]: Misleading to refer to this as "player" and to put it in this namespace.

**Review by dktapps — CHANGES_REQUESTED** [review]

**Comment by cooldogedev** [comment]
> What's the use case?

[Check out this example](https://github.com/cooldogepm/WDPELoginExtras). This plugin overrides the default Login packet handler to support WaterdogPE's login extras. However, it currently listens for all incoming packets, making the override process less than optimal in terms of convenience.

File: src/event/player/PlayerPacketHandlerChangeEvent.php (hunk 2 above)
Comment (cooldogedev) [reply src/event/player/PlayerPacketHandlerChangeEvent.php:30]: Without creating a new namespace it has to be either `event/player` or `event/server`

**Comment by dktapps** [comment]
WaterdogPE's design in general is less than optimal ...

**Comment by SenseiTarzan** [comment]
> WaterdogPE's design in general is less than optimal ...

I find this thing interesting because I have to do a hack to deactivate a handler because I have a synchronization system and it's me who manages the call that makes the setHandler(InGameHandler), WaterdogPE could make a new packets but they didn't and I find that a shame

**Comment by dktapps** [comment]
My gut reaction is that I don't like this change. The event's naming is problematic, but also I don't like the general concept of it.

Since this is primarily intended at supporting Waterdog, I feel like a more targeted approach to dealing with the extra JWT fields should be used.

**Comment by dktapps** [comment]
I think something that makes use of https://github.com/pmmp/netresearch-jsonmapper/blob/fdeac5f26eb7e61bdbab5e0948de8a71f3795102/src/JsonMapper.php#L112 would be better for WDPE than this approach. For example, if we added a `mixed[] $unknownProperties` element to the JSON models, and then added an undefined handler to collect all of them into one place, this would still allow the server to warn when there are unexpected new elements, while not preventing WDPE from working.

**Pull request closed without merging**